runtime/ops/fs.rs

```diff
@@ -820,58 +796,110 @@ async fn op_remove_async(
   .unwrap()
 }
 
-#[derive(Deserialize)]
-#[serde(rename_all = "camelCase")]
-pub struct CopyFileArgs {
-  from: String,
-  to: String,
-}
-
 #[op]
 fn op_copy_file_sync(
   state: &mut OpState,
-  args: CopyFileArgs,
+  from: String,
+  to: String,
 ) -> Result<(), AnyError> {
-  let from = PathBuf::from(&args.from);
-  let to = PathBuf::from(&args.to);
+  let from_path = PathBuf::from(&from);
+  let to_path = PathBuf::from(&to);
 
   let permissions = state.borrow_mut::<Permissions>();
-  permissions.read.check(&from)?;
-  permissions.write.check(&to)?;
+  permissions.read.check(&from_path)?;
+  permissions.write.check(&to_path)?;
 
-  debug!("op_copy_file_sync {} {}", from.display(), to.display());
   // On *nix, Rust reports non-existent `from` as ErrorKind::InvalidInput
   // See https://github.com/rust-lang/rust/issues/54800
   // Once the issue is resolved, we should remove this workaround.
-  if cfg!(unix) && !from.is_file() {
+  if cfg!(unix) && !from_path.is_file() {
     return Err(custom_error(
       "NotFound",
       format!(
         "File not found, copy '{}' -> '{}'",
-        from.display(),
-        to.display()
+        from_path.display(),
+        to_path.display()
       ),
     ));
   }
 
   let err_mapper = |err: Error| {
     Error::new(
       err.kind(),
-      format!("{}, copy '{}' -> '{}'", err, from.display(), to.display()),
+      format!(
+        "{}, copy '{}' -> '{}'",
+        err,
+        from_path.display(),
+        to_path.display()
+      ),
     )
   };
+
+  #[cfg(target_os = "macos")]
+  {
+    use libc::chmod;
+    use libc::clonefile;
+    use libc::stat;
+    use libc::unlink;
+    use std::ffi::CString;
+    use std::io::Read;
+
+    let from = CString::new(from).unwrap();
+    let to = CString::new(to).unwrap();
+
+    // SAFETY: `from` and `to` are valid C strings.
+    // std::fs::copy does open() + fcopyfile() on macOS. We try to use
+    // clonefile() instead, which is more efficient.
+    unsafe {
+      let mut st = std::mem::zeroed();
+      let ret = stat(from.as_ptr(), &mut st);
+      if ret != 0 {
+        return Err(err_mapper(Error::last_os_error()).into());
+      }
+
+      if st.st_size > 128 * 1024 {
+        // Try unlink.
+        let _ = unlink(to.as_ptr());
+        if clonefile(from.as_ptr(), to.as_ptr(), 0) == 0 {
+          // Preserve mode bits.
+          let _ = chmod(to.as_ptr(), st.st_mode);
```

Comment:
I am not sure that these error ignores here are safe. Also, this seems racy.

Reply:
i don't think it matters for the unlink as clonefile will fail if unlink fails but if we fail to set permissions correctly, then the whole operation should probably return an error or we should do some cleanup before falling back to normal copy (where it will likely fail again)?

tbh not sure i can think of a situation where we have managed to create the new file but aren't able to set the mode flags. :thinking:

Reply:
removed chmod to match Rust's io::copy behaviour.